packages/node/src/indexer/fetch.service.ts

```diff
@@ -0,0 +1,300 @@
+import { DynamicDsService } from './dynamic-ds.service';
+import { ProjectService } from './project.service';
+import {
+  blockToHeader,
+  UnfinalizedBlocksService,
+} from './unfinalizedBlocks.service';
+
+const logger = getLogger('fetch.service');
+
+const BLOCK_TIME_VARIANCE = 5000;
+
+const INTERVAL_PERCENT = 0.9;
+
+export function txFilterToQueryEntry(
+  filter: ConcordiumTransactionFilter,
+): DictionaryQueryEntry {
+  const conditions: DictionaryQueryCondition[] = [
+    {
+      field: 'type',
+      value: filter.type,
+      matcher: 'equalTo',
+    },
+  ];
+
+  if (filter.values !== undefined) {
+    const nested = {};
+
+    Object.keys(filter.values).map((key) => {
+      const value = filter.values[key];
+      setWith(nested, key, value);
+    });
+
+    conditions.push({
+      field: 'data',
+      value: nested as any, // Cast to any for compat with node core
+      matcher: 'contains',
+    });
+  }
+  return {
+    entity: 'transactions',
+    conditions: conditions,
+  };
+}
+
+export function txEventFilterToQueryEntry(
+  filter: ConcordiumTransactionEventFilter,
+): DictionaryQueryEntry {
+  const conditions: DictionaryQueryCondition[] = [
+    {
+      field: 'type',
+      value: filter.type,
+      matcher: 'equalTo',
+    },
+  ];
+
+  if (filter.values !== undefined) {
+    const nested = {};
+
+    Object.keys(filter.values).map((key) => {
+      const value = filter.values[key];
+      setWith(nested, key, value);
+    });
+
+    conditions.push({
+      field: 'data',
+      value: nested as any, // Cast to any for compat with node core
+      matcher: 'contains',
+    });
+  }
+  return {
+    entity: 'txEvents',
+    conditions: conditions,
+  };
+}
+
+export function speicalEventFilterToQueryEntry(
+  filter: ConcordiumSpecialEventFilter,
+): DictionaryQueryEntry {
+  const conditions: DictionaryQueryCondition[] = [
+    {
+      field: 'type',
+      value: filter.type,
+      matcher: 'equalTo',
+    },
+  ];
+
+  if (filter.values !== undefined) {
+    const nested = {};
+
+    Object.keys(filter.values).map((key) => {
+      const value = filter.values[key];
+      setWith(nested, key, value);
+    });
+
+    conditions.push({
+      field: 'data',
+      value: nested as any, // Cast to any for compat with node core
+      matcher: 'contains',
+    });
+  }
+  return {
+    entity: 'specialEvents',
+    conditions: conditions,
+  };
+}
+
+type GroupedConcordiumProjectDs = SubqlDatasource & {
+  groupedOptions?: SubqlConcordiumProcessorOptions[];
+};
+
+export function buildDictionaryQueryEntries(
+  dataSources: GroupedConcordiumProjectDs[],
+): DictionaryQueryEntry[] {
+  const queryEntries: DictionaryQueryEntry[] = [];
+
+  for (const ds of dataSources) {
+    for (const handler of ds.mapping.handlers) {
+      // No filters, cant use dictionary
+      if (!handler.filter) return [];
+
+      switch (handler.kind) {
+        case SubqlConcordiumHandlerKind.Block:
+          return [];
```

Comment:
I think there needs to be special cases for modulo and/or timestamp filters

Reply:
we don't query dictionary for block modulo and timestamp right?

Reply:
If there is a modulo filter we shouldn't disable the dictionary which is what this is doing.

I think it would be the same as this https://github.com/subquery/subql/blob/main/packages/node/src/indexer/fetch.service.ts#L151-L153